processing/src/main/java/org/apache/druid/segment/ObjectColumnSelector.java

```diff
@@ -19,75 +19,50 @@
+
 /**
- * This class is convenient for implementation of "object-sourcing" {@link ColumnValueSelector}s, it provides default
- * implementations for all {@link ColumnValueSelector}'s methods except {@link #getObject()} and {@link
- * #classOfObject()}.
- *
- * This class should appear ONLY in "extends" clause or anonymous class creation, but NOT in "user" code, where
- * {@link BaseObjectColumnValueSelector} must be used instead.
+ * Object based column selector.
  *
- * This is a class rather than an interface unlike {@link LongColumnSelector}, {@link FloatColumnSelector} and {@link
- * DoubleColumnSelector} solely in order to make {@link #isNull()} method final.
+ * This abstract class provides a {@link ColumnValueSelector} based on the methods of {@link BaseObjectColumnValueSelector}.
  */
 public abstract class ObjectColumnSelector<T> implements ColumnValueSelector<T>
 {
-  /**
-   * @deprecated This method is marked as deprecated in ObjectColumnSelector to minimize the probability of accidental
-   * calling. "Polymorphism" of ObjectColumnSelector should be used only when operating on {@link ColumnValueSelector}
-   * objects.
-   */
-  @Deprecated
+  private static final String COLUMN_IS_NULL_ERROR_MESSAGE = "Invalid usage pattern: method returning primitive called - but the column is null";
+
   @Override
-  public float getFloat()
+  public final float getFloat()
   {
     T value = getObject();
     if (value == null) {
-      return 0;
+      throw DruidException.defensive(COLUMN_IS_NULL_ERROR_MESSAGE);
     }
     return ((Number) value).floatValue();
   }
 
-  /**
-   * @deprecated This method is marked as deprecated in ObjectColumnSelector to minimize the probability of accidental
-   * calling. "Polymorphism" of ObjectColumnSelector should be used only when operating on {@link ColumnValueSelector}
-   * objects.
-   */
-  @Deprecated
   @Override
-  public double getDouble()
+  public final double getDouble()
   {
     T value = getObject();
     if (value == null) {
-      return 0;
+      throw DruidException.defensive(COLUMN_IS_NULL_ERROR_MESSAGE);
     }
     return ((Number) value).doubleValue();
   }
 
-  /**
-   * @deprecated This method is marked as deprecated in ObjectColumnSelector to minimize the probability of accidental
-   * calling. "Polymorphism" of ObjectColumnSelector should be used only when operating on {@link ColumnValueSelector}
-   * objects.
-   */
-  @Deprecated
   @Override
-  public long getLong()
+  public final long getLong()
   {
     T value = getObject();
     if (value == null) {
-      return 0;
+      throw DruidException.defensive(COLUMN_IS_NULL_ERROR_MESSAGE);
     }
     return ((Number) value).longValue();
   }
 
-  /**
-   * @deprecated This method is marked as deprecated in ObjectColumnSelector since it always returns false.
-   * There is no need to call this method.
-   */
-  @Deprecated
   @Override
   public final boolean isNull()
   {
-    return false;
+    return getObject() == null;
```

Comment:
while I guess this is harmless, nothing calling `getObject` should ever be using `isNull`, instead they should be checking that `getObject` returns null

if this change is fixing something then whatever it is fixing is probably wrong

Reply:
that said, i guess we could change the contract of `isNull` to always work, its just also sort of pointless since getObject can just return the null (while getLong/getDouble/getFloat cannot since primitives), so why call two methods when you only need to call one

Reply:
I think this method should be either implemented correctly or throw an exception that its not supported - that will make sure that there won't be correctness issues if somehow it gets called

it is called from places like [this](https://github.com/apache/druid/blob/93cd63864581b16d6b9407723541616a7465da68/processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregator.java#L60) - how should we fix that?

I don't understand clearly what you mean by "changing the contract of isNull" ; I believe that's defined [here](https://github.com/apache/druid/blob/93cd63864581b16d6b9407723541616a7465da68/processing/src/main/java/org/apache/druid/segment/BaseNullableColumnValueSelector.java#L31-L42); and returning `false` have kinda violated it

Reply:
>I think this method should be either implemented correctly or throw an exception that its not supported - that will make sure that there won't be correctness issues if somehow it gets called

That is what i was getting at, if `getLong`/`getFloat`/`getDouble` are going to throw exceptions, then this should too, since it should only be called if you are planning on calling one of those methods.


>it is called from places like this - how should we fix that?

`NullableNumericAggregator` is _only_ for things that are going to be calling `getLong`/`getFloat`/`getDouble`, which are the only things that are supposed to call `isNull`, because those methods return primitives which cannot be null. That aggregator should never be using a selector derived from this one, since it is designed to aggregate primitive numeric values.

>I don't understand clearly what you mean by "changing the contract of isNull" ; I believe that's defined [here](https://github.com/apache/druid/blob/93cd63864581b16d6b9407723541616a7465da68/processing/src/main/java/org/apache/druid/segment/BaseNullableColumnValueSelector.java#L31-L42); and returning false have kinda violated it

This does not violate it, because no one should _ever_ call `getLong`/`getFloat`/`getDouble` on that value selector, which means they should also not be using `isNull`, and should instead just call `getObject` and see if the result is null. Calling `isNull` if you intend to call `getObject` is violating the contract.

Reply:
> That is what i was getting at, if getLong/getFloat/getDouble are going to throw exceptions, then this should too, since it should only be called if you are planning on calling one of those methods.

I think it should throw an exception because `null` can't be represented as a primitive; it could highlight that you should call `isNull` first - I can revert that change; but I think those default values may only help to hide bugs....

> [...] NullableNumericAggregator is only for things that are going to be calling [...]

that sounds rather odd - there are no interfaces enforcing that contract; it only requires that the `selector` implement `BaseNullableColumnValueSelector` ; and because of that object which implement that should work with it

This `ObjectColumnSelector` class provides `getLong` and  friends based on `getObject` which is exactly how its services are used [here](https://github.com/apache/druid/blob/0516d0dae432058848806e49e71ac73333bf751c/processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java#L554-L575)

>This does not violate it, because no one should ever call

"no one should ever call" : that can't be enforced; it
providing a `return false` implementation is wrong - and a violation of the contract in this case.

> Calling isNull if you intend to call getObject is violating the contract.
now you lost me...then why implement `ColumnValueSelector` at all?  or not just throw exceptions from all of them?
this seems like some kind of design issue...

I think a possible alternate way to fix this could be to *not* implement `isNull` in `ObjectColumnSelector` and force classes extending it to provide their own implementation - what do you think ?

Reply:
I think https://github.com/apache/druid/blob/0516d0dae432058848806e49e71ac73333bf751c/processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java#L554-L575 is mistake, and this should not be making an `ObjectColumnSelector` and instead should implement `ColumnValueSelector` directly.

As far as I can tell, everything else using `ObjectColumnSelector` is truly something that will only call getObject like complex types of various sorts.

Reply:
I believe there are two ways to provide a `ColumnValueSelector` for `Object` -t:
* one which doesn't want to be treated like something which could answer a call like `getLong()` or `isNull()` as it wants to enforce that people will use the `getObject()` call
* another one which provides the `getLong` and other methods based on a backing `getObject` which *could* be even a primitive

The old `ObjectColumnSelector` was kinda tring to do both of the above; in the latest changeset I've modified it to simply throw an Exception for every `getX` call to prevent incorrect usage - I needed to make some changes to some tests as they were asserting on the `isNull` method (which was implemented by `return false`).

I've added ta second class which does the 2nd kind of thing named `ObjectBasedColumnSelector` - and its used in the windowing stuff and a few other places; I don't see much value of copy-pasting that implementation to every place as if you have an array of objects. If someone might want to call `getLong()` on it there is not really a alternate ways...if this ends up being slow or something - and identified as a performance bottleneck...we could roll `ArrayBasedTypeColumnSelector` classes via a strategy pattern and done...but right now I doubt that around windowing this would be the 1st bottleneck :D